README
giix Readme
===========

giix version 1.10

giix official website:    http://giix.org/

giix on...
Yii Extension Repository: http://www.yiiframework.com/extension/giix/
Yii Forum:                http://www.yiiframework.com/forum/index.php?/topic/13154-giix-%E2%80%94-gii-extended/
Google Code:              http://giix.googlecode.com/
Ohloh:                    http://www.ohloh.net/p/giix

giix is gii Extended, a code generator for Yii PHP framework.
giix is inspired and based on gii-template-collection (gtc), by Herbert Maschke (thyseus).

giix is free software and is dual-licensed under the terms of the new BSD License and under the terms of GNU GPL v3. See the LICENSE file.

ACKNOWLEDGEMENTS
----------------

giix is inspired and uses code from Yii PHP framework and gii-template-collection.
Many thanks to Qiang Xue, Herbert Maschke and the contributors of these software.

Also, thanks for the huge help from:
Thiago Talma (Athos)
Gustavo Salomé Silva (Gustavo)

FEATURES
--------

giix extends Yii's gii by providing:
- Proper handling of related model attributes, rendering appropriate form fields based on relation type.
- More support for HAS_MANY and MANY_MANY relations.
- Native support for saving MANY_MANY relations with the new method GxActiveRecord::saveWithRelated.
- Native support for saving multiple (related or not) records with the new method GxActiveRecord::saveMultiple.
- Automatic string representation for a model via GxActiveRecord::representingColumn() and GxActiveRecord::__toString().
- Better i18n support and easier to maintain: support to active record labels. You change the label once in the model and it is automatically updated in the views. This is especially useful for plurals.
- Out-of-the box i18n support by using Yii::t().
- Bundled translations (pending support on Yii core as of Yii 1.1.8. See <http://code.google.com/p/yii/issues/detail?id=2624> for details).
- Appropriate form fields are rendered based on model attribute/table field data type.
- Separated model and basemodel. Basemodel can be regenerated without overwriting your code in the model.
- Smart methods can query your database for just the needed data (usually the primary key and the field with the string representantion), avoiding manual setup or a "select *".
- Extensive use of default method parameters. Appropriate data is found automatically.
- Some (incipient) support for tables with composite primary keys.
- Generated code is completely free from styling and formatting. Your CSS controls the presentation.
- Generated code is (almost) free of comments.
- Well documented and commented source code. Ohloh says that 40% of the lines in the code are comments! You can understand and modify anything you want.

And a lot more! Read the CHANGELOG file and the (richly commented) source code to fully leverage giix's power.

Some of these features come from gtc.

WARNINGS
--------

giix is not fully tested now, so please test your application and be careful using it.
giix is still in development. Some changes may break backwards compatibility.

Are you upgrading? Please don't forget to read the UPGRADE file for instructions!

REQUIREMENTS
------------

Yii 1.1.9+ is required.

giix should run with:
- Yii 1.1.8+ (tested with revision 3364 + patch*)
- Linux
- Apache 2.2 or nginx 0.8+
- MySQL 5.x
- PHP 5.2+

giix may run on other setups, but there is no official suppport for them.
There is no planned official support for other setups, including Windows/IIS, XAMPP/etc or other DBMS.

* You can use giix with Yii 1.1.8 if you patch your Yii installation with the patch provided in the issue 2624:
<http://code.google.com/p/yii/issues/detail?id=2624>.

INSTALLATION AND UPGRADING
--------------------------

Please see INSTALL and UPGRADE files for instructions.

USING
-----

After installing, you will see the giix generators on the gii page.
The giix generators' names start with "giix".
Read the gii doumentation on how to use it
<http://www.yiiframework.com/doc/api/GiiModule>.

If you want to use the GxActiveRecord::saveWithRelated method, you must first generate
the pivot model for the cross-reference table of the relationship.
The pivot model class is the active record between two other active records that have
a MANY_MANY relationship.
If you have already generated models for every table in your schema, you are ready to
use the method to save your MANY_MANY relations.
If you use the support for saving MANY_MANY relations as generated, you will need to
manually write code to support scenarios. This is an important security concern.

DONATE
------

Go to <http://giix.org/> and donate via Flattr or via Paypal.

CHANGELOG
giix Change Log
===============

Version 1.10, to be released
----------------------------
Chg: Better directory organization. Now giix have its own root directory.
Bug: Fixes issue 23: Removed undefined variables from GxActiveRecord::saveRelated.
Chg: Removed the message translation from the generated view files.
Enh: Added a message category to the Yii::t calls.
Enh: Added the configuration file for generating message translations.
Enh: Added instructions and details about the patch and messages directory configuration.
Bug: GxActiveRecord::pkSeparator and GxActiveRecord::repColumnsSeparator shouldn't be static.
Bug: Now generating the proper label for the relation in the view "_view".

Version 1.9, 2011-07-06
----------------------------
Bug: GxActiveRecord::getRelationLabel supported only relations and (table-based) attributes. Support for class properties has been added.
Chg: The method attributeLabels in the generated base model now have the FK fields set to null by default.
Chg: The method attributeLabels in the generated base model now have the relations of the model. They are set to null by default.
Chg: The method GxActiveRecord::label doesn't have the second parameter "ucwords" anymore. The same happens when it is overriden on the generated base model.
Chg: The method GxActiveRecord::getRelationLabel doesn't have the third parameter "ucwords" anymore.
Chg: Now the method GxActiveRecord::getRelationLabel gives priority to the explicitly defined label in the model.

Version 1.8, 2011-06-20
----------------------------
Enh: Added support to active record labels.

Version 1.7, 2011-06-02
----------------------------
Enh: Updated the code from Yii to version 1.1.7 (r3135).
Enh: Updated the code from Yii in the default template files and generator's index views to version 1.1.7 (r3135).
Bug: Fixed bug #10. Added proper support to 'uncheckValue'.
Bug: Fixes issue #11. Now the generated model::search method uses the current instance of the model and allows the use of custom criteria.

Version 1.6, 2011-05-09
----------------------------
Bug: Wrong pivot class name was generated when there were table prefixes.
Enh: Better support to boolean values: formatted display in view "view".
Bug: Added proper parameter checks in GxActiveRecord::__toString.
Chg: Raised Yii version in "requirements".
Enh: Added support for entering PK values. See http://code.google.com/p/yii/issues/detail?id=1770.
Chg: New default value for GxActiveRecord::saveMultiple: 'detectRelations' option is now false by default.
Enh: Added GxActiveRecord::fillPkColumnNames.
Enh: Now GxController::loadModel supports composite PK, has autodetection features and can be used with any attribute (column).

Version 1.5, 2011-05-06
----------------------------
Chg: GxController extends again from Controller (reverted from version 1.2).
Enh: Now GxController::performAjaxValidation supports multiple models.
Bug: Removed the code for saving the main model and committing the transaction from the loop in GxActiveRecord::saveWithRelated.
Bug: Fixed a bug where the function would return instead of continue iterating in the loop in GxActiveRecord::saveWithRelated.
Bug: Do not swallow exceptions anymore in GxActiveRecord::saveWithRelated.
Bug: Now more code is wrapped in the try-catch block to protect the integrity of the transaction in GxActiveRecord::saveWithRelated.
Bug: Fixed issue 5.
Enh: Introduced GxActiveRecord::saveMultiple.
Chg: Removed the "cancel" button from the _form view.

Version 1.4, 2010-12-30
----------------------------
Bug: Saving related MANY_MANY relations could delete or insert wrong data since the check on current relations was not necessarily made against the database.

Version 1.3, 2010-12-23
----------------------------
New: Added native support for saving MANY_MANY relations with the new method GxActiveRecord::saveWithRelated.
New: Added method GxActiveRecord::pivotModels.
New: Added automatic detection of HAS_MANY and MANY_MANY relations in the POST with the new method GxController::getRelatedData.
New: The generated model now have the pivot class names in its pivotModels method.
New: The generated controller now have support for saving MANY_MANY relations automatically.

Version 1.2, 2010-12-14
----------------------------
Enh: More inline styling and formatting is removed from the generated code.
Enh: Now the breadcrumb will contain the string representation of the model instance instead of the primary key value, when appropriate.
Enh: Now the page heading will contain the string representation of the model instance instead of the primary key value, when appropriate.
Bug: Fixed a bug were the action name were translated in the breadcrumb link.
Chg: Now GxController extends directly from CController. Class Controller is no longer necessary.
Enh: I18n support added to some messages.

Version 1.1.1, 2010-12-10
----------------------------
Enh: All inline styling and formatting is removed from the generated code.
Bug: GxActiveForm::enableAjaxValidation value was passed as a string instead of as a boolean in _form.php, causing it to ever evaluate to true. This caused a new record to be added in every ajax validation post (There weren't a call to GxController::performAjaxValidation).
Chg: Removed support for enum fields.
Bug: Now all the output data is HTML encoded.
New: Added method GxHtml::encodeEx.

Version 1.0, 2010-11-12
----------------------------
Initial release, based on gii-template-collection (gtc).
Updated up to gtc r107 and yii/gii 1.1.3 r2247.

generators/giixModel/GiixModelCode.php
<?php

/**
 * GiixModelCode class file.
 *
 * @author Rodrigo Coelho <rodrigo@giix.org>
 * @link http://giix.org/
 * @copyright Copyright &copy; 2010-2011 Rodrigo Coelho
 * @license http://giix.org/license/ New BSD License
 */
Yii::import('system.gii.generators.model.ModelCode');
Yii::import('ext.giix.helpers.*');

/**
 * GiixModelCode is the model for giix model generator.
 *
 * @author Rodrigo Coelho <rodrigo@giix.org>
 * @package giix.generators.giixModel
 */
class GiixModelCode extends ModelCode {

	/**
	 * @var string The (base) model base class name.
	 */
	public $baseClass = 'GxActiveRecord';
	/**
	 * @var string The path of the base model.
	 */
	public $baseModelPath;
	/**
	 * @var string The base model class name.
	 */
	public $baseModelClass;

	/**
	 * Prepares the code files to be generated.
	 * #MethodTracker
	 * This method is based on {@link ModelCode::prepare}, from version 1.1.7 (r3135). Changes:
	 * <ul>
	 * <li>Generates the base model.</li>
	 * <li>Provides the representing column for the table.</li>
	 * <li>Provides the pivot class names for MANY_MANY relations.</li>
	 * </ul>
	 */
	public function prepare() {
		if (($pos = strrpos($this->tableName, '.')) !== false) {
			$schema = substr($this->tableName, 0, $pos);
			$tableName = substr($this->tableName, $pos + 1);
		} else {
			$schema = '';
			$tableName = $this->tableName;
		}
		if ($tableName[strlen($tableName) - 1] === '*') {
			$tables = Yii::app()->db->schema->getTables($schema);
			if ($this->tablePrefix != '') {
				foreach ($tables as $i => $table) {
					if (strpos($table->name, $this->tablePrefix) !== 0)
						unset($tables[$i]);
				}
			}
		}
		else
			$tables=array($this->getTableSchema($this->tableName));

		$this->files = array();
		$templatePath = $this->templatePath;

		$this->relations = $this->generateRelations();

		foreach ($tables as $table) {
			$tableName = $this->removePrefix($table->name);
			$className = $this->generateClassName($table->name);

			// Generate the pivot model data.
			$pivotModels = array();
			if (isset($this->relations[$className])) {
				foreach ($this->relations[$className] as $relationName => $relationData) {
					if (preg_match('/^array\(self::MANY_MANY,.*?,\s*\'(.+?)\(/', $relationData, $matches)) {
						// Clean the table name if needed.
						$pivotTableName = str_replace(array('{', '}'), '', $matches[1]);
						$pivotModels[$relationName] = $this->generateClassName($pivotTableName);
					}
				}
			}

			$params = array(
				'tableName' => $schema === '' ? $tableName : $schema . '.' . $tableName,
				'modelClass' => $className,
				'columns' => $table->columns,
				'labels' => $this->generateLabelsEx($table, $className),
				'rules' => $this->generateRules($table),
				'relations' => isset($this->relations[$className]) ? $this->relations[$className] : array(),
				'representingColumn' => $this->getRepresentingColumn($table), // The representing column for the table.
				'pivotModels' => $pivotModels, // The pivot models.
			);
			// Setup base model information.
			$this->baseModelPath = $this->modelPath . '._base';
			$this->baseModelClass = 'Base' . $className;
			// Generate the model.
			$this->files[] = new CCodeFile(
							Yii::getPathOfAlias($this->modelPath . '.' . $className) . '.php',
							$this->render($templatePath . DIRECTORY_SEPARATOR . 'model.php', $params)
			);
			// Generate the base model.
			$this->files[] = new CCodeFile(
							Yii::getPathOfAlias($this->baseModelPath . '.' . $this->baseModelClass) . '.php',
							$this->render($templatePath . DIRECTORY_SEPARATOR . '_base' . DIRECTORY_SEPARATOR . 'basemodel.php', $params)
			);
		}
	}

	/**
	 * Lists the template files.
	 * #MethodTracker
	 * This method is based on {@link ModelCode::requiredTemplates}, from version 1.1.7 (r3135). Changes:
	 * <ul>
	 * <li>Includes the base model.</li>
	 * </ul>
	 * @return array A list of required template filenames.
	 */
	public function requiredTemplates() {
		return array(
			'model.php',
			'_base' . DIRECTORY_SEPARATOR . 'basemodel.php',
		);
	}

	/**
	 * Generates the labels for the table fields and relations.
	 * By default, the labels for the FK fields and for the relations is null. This
	 * will cause them to be represented by the related model label.
	 * #MethodTracker
	 * This method is based on {@link ModelCode::generateLabels}, from version 1.1.7 (r3135). Changes:
	 * <ul>
	 * <li>Default label for FKs is null.</li>
	 * <li>Creates entries for the relations. The default label is null.</li>
	 * </ul>
	 * @param CDbTableSchema $table The table definition.
	 * @param string $className The model class name.
	 * @return array The labels.
	 * @see GxActiveRecord::label
	 * @see GxActiveRecord::getRelationLabel
	 */
	public function generateLabelsEx($table, $className) {
		$labels = array();
		// For the fields.
		foreach ($table->columns as $column) {
			if ($column->isForeignKey) {
				$label = null;
			} else {
				$label = ucwords(trim(strtolower(str_replace(array('-', '_'), ' ', preg_replace('/(?<![A-Z])[A-Z]/', ' \0', $column->name)))));
				$label = preg_replace('/\s+/', ' ', $label);

				if (strcasecmp(substr($label, -3), ' id') === 0)
					$label = substr($label, 0, -3);
				if ($label === 'Id')
					$label = 'ID';

				$label = "Yii::t('app', '{$label}')";
			}
			$labels[$column->name] = $label;
		}
		// For the relations.
		$relations = $this->getRelationsData($className);
		if (isset($relations)) {
			foreach (array_keys($relations) as $relationName) {
				$labels[$relationName] = null;
			}
		}

		return $labels;
	}

	/**
	 * Generates the rules for table fields.
	 * #MethodTracker
	 * This method overrides {@link ModelCode::generateRules}, from version 1.1.7 (r3135). Changes:
	 * <ul>
	 * <li>Adds the rule to fill empty attributes with null.</li>
	 * </ul>
	 * @param CDbTableSchema $table The table definition.
	 * @return array The rules for the table.
	 */
	public function generateRules($table) {
		$rules = array();
		$null = array();
		foreach ($table->columns as $column) {
			if ($column->autoIncrement)
				continue;
			if (!(!$column->allowNull && $column->defaultValue === null))
				$null[] = $column->name;
		}
		if ($null !== array())
			$rules[] = "array('" . implode(', ', $null) . "', 'default', 'setOnEmpty' => true, 'value' => null)";

		return array_merge(parent::generateRules($table), $rules);
	}

	/**
	 * Selects the representing column of the table.
	 * The "representingColumn" method is the responsible for the
	 * string representation of the model instance.
	 * @param CDbTableSchema $table The table definition.
	 * @return string|array The name of the column as a string or the names of the columns as an array.
	 * @see GxActiveRecord::representingColumn
	 * @see GxActiveRecord::__toString
	 */
	protected function getRepresentingColumn($table) {
		$columns = $table->columns;
		// If this is not a MANY_MANY pivot table
		if (!$this->isRelationTable($table)) {
			// First we look for a string, not null, not pk, not fk column, not original number on db.
			foreach ($columns as $name => $column) {
				if ($column->type === 'string' && !$column->allowNull && !$column->isPrimaryKey && !$column->isForeignKey && stripos($column->dbType, 'int') === false)
					return $name;
			}
			// Then a string, not null, not fk column, not original number on db.
			foreach ($columns as $name => $column) {
				if ($column->type === 'string' && !$column->allowNull && !$column->isForeignKey && stripos($column->dbType, 'int') === false)
					return $name;
			}
			// Then the first string column, not original number on db.
			foreach ($columns as $name => $column) {
				if ($column->type === 'string' && stripos($column->dbType, 'int') === false)
					return $name;
			}
		} // If the appropriate column was not found or if this is a MANY_MANY pivot table.
		// Then the pk column(s).
		$pk = $table->primaryKey;
		if ($pk !== null) {
			if (is_array($pk))
				return $pk;
			else
				return (string) $pk;
		}
		// Then the first column.
		return reset($columns)->name;
	}

	/**
	 * Finds the related class of the specified column.
	 * @param string $className The model class name.
	 * @param CDbColumnSchema $column The column.
	 * @return string The related class name. Or null if no matching relation was found.
	 */
	public function findRelatedClass($className, $column) {
		if (!$column->isForeignKey)
			return null;

		$relations = $this->getRelationsData($className);

		foreach ($relations as $relation) {
			// Must be BELONGS_TO.
			if (($relation[0] === GxActiveRecord::BELONGS_TO) && ($relation[3] === $column->name))
				return $relation[1];
		}
		// None found.
		return null;
	}

	/**
	 * Finds the relation data for all the relations of the specified model class.
	 * @param string $className The model class name.
	 * @return array An array of arrays with the relation data.
	 * The array will have one array for each relation.
	 * The key is the relation name. There are 5 values:
	 * 0: the relation type,
	 * 1: the related active record class name,
	 * 2: the joining (pivot) table (note: it may come with curly braces) (if the relation is a MANY_MANY, else null),
	 * 3: the local FK (if the relation is a BELONGS_TO or a MANY_MANY, else null),
	 * 4: the remote FK (if the relation is a HAS_ONE, a HAS_MANY or a MANY_MANY, else null).
	 * Or null if the model has no relations.
	 */
	public function getRelationsData($className) {
		if (!empty($this->relations))
			$relations = $this->relations;
		else
			$relations = $this->generateRelations();

		if (!isset($relations[$className]))
			return null;

		$result = array();
		foreach ($relations[$className] as $relationName => $relationData) {
			$result[$relationName] = $this->getRelationData($className, $relationName, $relations);
		}
		return $result;
	}

	/**
	 * Finds the relation data of the specified relation name.
	 * @param string $className The model class name.
	 * @param string $relationName The relation name.
	 * @param array $relations An array of relations for the models
	 * in the format returned by {@link ModelCode::generateRelations}. Optional.
	 * @return array The relation data. The array will have 3 values:
	 * 0: the relation type,
	 * 1: the related active record class name,
	 * 2: the joining (pivot) table (note: it may come with curly braces) (if the relation is a MANY_MANY, else null),
	 * 3: the local FK (if the relation is a BELONGS_TO or a MANY_MANY, else null),
	 * 4: the remote FK (if the relation is a HAS_ONE, a HAS_MANY or a MANY_MANY, else null).
	 * Or null if no matching relation was found.
	 */
	public function getRelationData($className, $relationName, $relations = array()) {
		if (empty($relations)) {
			if (!empty($this->relations))
				$relations = $this->relations;
			else
				$relations = $this->generateRelations();
		}

		if (isset($relations[$className]) && isset($relations[$className][$relationName]))
			$relation = $relations[$className][$relationName];
		else
			return null;

		$relationData = array();
		if (preg_match("/^array\(([\w:]+?),\s?'(\w+)',\s?'([\w\s\(\),]+?)'\)$/", $relation, $matches_base)) {
			$relationData[1] = $matches_base[2]; // the related active record class name

			switch ($matches_base[1]) {
				case 'self::BELONGS_TO':
					$relationData[0] = GxActiveRecord::BELONGS_TO; // the relation type
					$relationData[2] = null;
					$relationData[3] = $matches_base[3]; // the local FK
					$relationData[4] = null;
					break;
				case 'self::HAS_ONE':
					$relationData[0] = GxActiveRecord::HAS_ONE; // the relation type
					$relationData[2] = null;
					$relationData[3] = null;
					$relationData[4] = $matches_base[3]; // the remote FK
					break;
				case 'self::HAS_MANY':
					$relationData[0] = GxActiveRecord::HAS_MANY; // the relation type
					$relationData[2] = null;
					$relationData[3] = null;
					$relationData[4] = $matches_base[3]; // the remote FK
					break;
				case 'self::MANY_MANY':
					if (preg_match("/^((?:{{)?\w+(?:}})?)\((\w+),\s?(\w+)\)$/", $matches_base[3], $matches_manymany)) {
						$relationData[0] = GxActiveRecord::MANY_MANY; // the relation type
						$relationData[2] = $matches_manymany[1]; // the joining (pivot) table
						$relationData[3] = $matches_manymany[2]; // the local FK
						$relationData[4] = $matches_manymany[3]; // the remote FK
					}
					break;
			}

			return $relationData;
		} else
			return null;
	}

	/**
	 * Returns the message to be displayed when the newly generated code is saved successfully.
	 * #MethodTracker
	 * This method overrides {@link CCodeModel::successMessage}, from version 1.1.7 (r3135). Changes:
	 * <ul>
	 * <li>Custom giix success message.</li>
	 * </ul>
	 * @return string The message to be displayed when the newly generated code is saved successfully.
	 */
	public function successMessage() {
		return <<<EOM
<p><strong>Sweet!</strong></p>
<ul style="list-style-type: none; padding-left: 0;">
	<li><img src="http://giix.org/icons/love.png"> Show how you love giix on <a href="http://www.yiiframework.com/forum/index.php?/topic/13154-giix-%E2%80%94-gii-extended/">the forum</a> and on its <a href="http://www.yiiframework.com/extension/giix">extension page</a></li>
	<li><img src="http://giix.org/icons/vote.png"> Upvote <a href="http://www.yiiframework.com/extension/giix">giix</a></li>
	<li><img src="http://giix.org/icons/powered.png"> Show everybody that you are using giix in <a href="http://www.yiiframework.com/forum/index.php?/topic/19226-powered-by-giix/">Powered by giix</a></li>
	<li><img src="http://giix.org/icons/donate.png"> <a href="http://giix.org/">Donate</a></li>
</ul>
<p style="margin: 2px 0; position: relative; text-align: right; top: -15px; color: #668866;">icons by <a href="http://www.famfamfam.com/lab/icons/silk/" style="color: #668866;">famfamfam.com</a></p>
EOM;
	}

}